vaadin-dialog-flow-parent/vaadin-dialog-flow/src/main/java/com/vaadin/flow/component/dialog/Dialog.java

```diff
@@ -528,6 +531,165 @@ public boolean isResizable() {
+    public DialogHeader getHeader() {
+        if (this.dialogHeader == null) {
+            this.dialogHeader = new DialogHeader(this);
+        }
+        return this.dialogHeader;
+    }
+
+    /**
+     * Gets the object from which components can be added or removed from the
+     * dialog footer area. The footer is displayed only if there's at least one
+     * component added with {@link DialogHeaderFooter#add(Component)}.
+     *
+     * @return the header object
+     */
+    public DialogFooter getFooter() {
+        if (this.dialogFooter == null) {
+            this.dialogFooter = new DialogFooter(this);
+        }
+        return this.dialogFooter;
+    }
+
+    /**
+     * Class for adding and removing components to the header part of a dialog.
+     */
+    final public static class DialogHeader extends DialogHeaderFooter {
+        private DialogHeader(Dialog dialog) {
+            super("headerRenderer", dialog);
+        }
+    }
+
+    /**
+     * Class for adding and removing components to the header part of a dialog.
+     */
+    final public static class DialogFooter extends DialogHeaderFooter {
+        private DialogFooter(Dialog dialog) {
+            super("footerRenderer", dialog);
+            root.getStyle().set("flex", "1");
+            root.getStyle().set("justify-content", "flex-end");
+        }
+    }
+
+    /**
+     * This class defines the common behavior for adding/removing components to
+     * the header and footer parts. It also creates the root element where the
+     * components will be attached to as well as the renderer function used by
+     * the dialog.
+     */
+    abstract static class DialogHeaderFooter implements Serializable {
+        protected final Element root;
+        private final String rendererFunction;
+        private final Component dialog;
+        boolean rendererCreated = false;
+
+        protected DialogHeaderFooter(String rendererFunction,
+                Component dialog) {
+            this.rendererFunction = rendererFunction;
+            this.dialog = dialog;
+            root = new Element("div");
+            root.getStyle().set("display", "flex");
+        }
+
+        /**
+         * Adds the component to the container.
+         *
+         * @param component
+         *            the component to be added.
+         */
+        public void add(Component component) {
+            root.appendChild(component.getElement());
+            if (!isRendererCreated()) {
+                initRenderer();
+            }
+        }
+
+        /**
+         * Removes the component from the container.
+         *
+         * <p>
+         * Note that the component needs to be removed from this method in order
+         * to guarantee the correct state of the component.
+         *
+         * @param component
+         *            the component to be removed.
+         */
+        public void remove(Component component) {
+            root.removeChild(component.getElement());
```

Comment:
Le't add a check whether `root` is a parent of the given `component` or not. Otherwise there will be an error (it can happen if you try to remove an element that has been already removed):

```
[qtp754836900-40] ERROR com.vaadin.flow.server.DefaultErrorHandler -
java.lang.IllegalArgumentException: Child should have this element as a parent
	at com.vaadin.flow.dom.Node.ensureChildHasParent(Node.java:483)
	at com.vaadin.flow.dom.Node.removeChild(Node.java:376)
	at com.vaadin.flow.component.dialog.Dialog$DialogHeaderFooter.remove(Dialog.java:634)
	at com.vaadin.flow.component.dialog.Dialog$DialogHeader.remove(Dialog.java:582)
```

Reply:
Done.